File: apps/docker-server/cmd/renewlet/account_security_key.go
package main

import (
	"crypto/hkdf"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sync"

	"github.com/pocketbase/pocketbase/core"
)

const (
	accountSecurityKeyVersion = 1
	accountSecurityKeyBytes   = 32
	accountSecurityKeySalt    = "renewlet:account-security:v1"
)

var (
	errAccountSecurityKeyUnavailable = errors.New("account security key is unavailable")
	accountSecurityKeyCacheMu        sync.Mutex
	accountSecurityKeyCache          = map[string]*accountSecurityKeyRing{}
)

type accountSecurityKeyRing struct {
	totpSeed          []byte
	sharingCredential []byte
	recoveryCode      []byte
	mfaTicket         []byte
	passkeyChallenge  []byte
}

type accountSecurityKeyFile struct {
	Version int    `json:"version"`
	Key     string `json:"key"`
}

// accountSecurityKeyRingForApp 缓存安装级派生 key ring；缓存 key 用 DataDir 隔离测试/实例，不缓存用户或请求态。
func accountSecurityKeyRingForApp(app core.App) (*accountSecurityKeyRing, error) {
	dataDir := app.DataDir()
	if dataDir == "" {
		return nil, errAccountSecurityKeyUnavailable
	}
	accountSecurityKeyCacheMu.Lock()
	defer accountSecurityKeyCacheMu.Unlock()
	if ring := accountSecurityKeyCache[dataDir]; ring != nil {
		return ring, nil
	}
	master, err := loadOrCreateAccountSecurityMasterKey(dataDir)
	if err != nil {
		return nil, err
	}
	ring, err := deriveAccountSecurityKeyRing(master)
	if err != nil {
		return nil, err
	}
	accountSecurityKeyCache[dataDir] = ring
	return ring, nil
}

func loadOrCreateAccountSecurityMasterKey(dataDir string) ([]byte, error) {
	path := accountSecurityKeyPath(dataDir)
	master, err := readAccountSecurityMasterKey(path)
	if err == nil || !errors.Is(err, os.ErrNotExist) {
		return master, err
	}
	master = make([]byte, accountSecurityKeyBytes)
	if _, err := rand.Read(master); err != nil {
		return nil, err
	}
	if err := os.MkdirAll(filepath.Dir(path), 0o700); err != nil {
		return nil, err
	}
	if err := os.Chmod(filepath.Dir(path), 0o700); err != nil {
		return nil, err
	}
	if err := writeAccountSecurityMasterKey(path, master); err != nil {
		if errors.Is(err, os.ErrExist) {
			return readAccountSecurityMasterKey(path)
		}
		return nil, err
	}
	return master, nil
}

func readAccountSecurityMasterKey(path string) ([]byte, error) {
	content, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var stored accountSecurityKeyFile
	if err := json.Unmarshal(content, &stored); err != nil {
		return nil, fmt.Errorf("%w: invalid account security key file", errAccountSecurityKeyUnavailable)
	}
	if stored.Version != accountSecurityKeyVersion {
		return nil, fmt.Errorf("%w: unsupported account security key version", errAccountSecurityKeyUnavailable)
	}
	master, err := base64.RawURLEncoding.DecodeString(stored.Key)
	if err != nil || len(master) != accountSecurityKeyBytes {
		return nil, fmt.Errorf("%w: invalid account security key material", errAccountSecurityKeyUnavailable)
	}
	return master, nil
}

func writeAccountSecurityMasterKey(path string, master []byte) error {
	payload, err := json.Marshal(accountSecurityKeyFile{
		Version: accountSecurityKeyVersion,
		Key:     base64.RawURLEncoding.EncodeToString(master),
	})
	if err != nil {
		return err
	}
	// 安装级账号安全密钥不进 env，不进数据库；0600 文件让 Docker 零配置同时避免普通导出泄漏 pepper。
	file, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0o600)
	if err != nil {
		return err
	}
	if _, err := file.Write(append(payload, '\n')); err != nil {
		_ = file.Close()
		return err
	}
	return file.Close()
}

func deriveAccountSecurityKeyRing(master []byte) (*accountSecurityKeyRing, error) {
	// HKDF Extract 只做一次，四个用途靠 info 分域；缓存命中后登录热路径不会再触碰文件或重复派生。
	prk, err := hkdf.Extract(sha256.New, master, []byte(accountSecurityKeySalt))
	if err != nil {
		return nil, err
	}
	totpSeed, err := deriveAccountSecurityKey(prk, "totp-seed-aes-gcm")
	if err != nil {
		return nil, err
	}
	sharingCredential, err := deriveAccountSecurityKey(prk, "sharing-credential-aes-gcm")
	if err != nil {
		return nil, err
	}
	recoveryCode, err := deriveAccountSecurityKey(prk, "recovery-code-hmac")
	if err != nil {
		return nil, err
	}
	mfaTicket, err := deriveAccountSecurityKey(prk, "mfa-ticket-hmac")
	if err != nil {
		return nil, err
	}
	passkeyChallenge, err := deriveAccountSecurityKey(prk, "passkey-challenge-hmac")
	if err != nil {
		return nil, err
	}
	return &accountSecurityKeyRing{
		totpSeed:          totpSeed,
		sharingCredential: sharingCredential,
		recoveryCode:      recoveryCode,
		mfaTicket:         mfaTicket,
		passkeyChallenge:  passkeyChallenge,
	}, nil
}

func deriveAccountSecurityKey(prk []byte, info string) ([]byte, error) {
	return hkdf.Expand(sha256.New, prk, info, accountSecurityKeyBytes)
}

func accountSecurityKeyPath(dataDir string) string {
	return filepath.Join(dataDir, "system", "account-security-key.v1.json")
}

func resetAccountSecurityKeyRingCacheForTest() {
	accountSecurityKeyCacheMu.Lock()
	defer accountSecurityKeyCacheMu.Unlock()
	accountSecurityKeyCache = map[string]*accountSecurityKeyRing{}
}

File: apps/worker/src/account-security-key.ts
import type { Env } from "./types";
import { arrayBufferFromBytes, base64Url, base64UrlToArrayBuffer } from "./encoding";

const ACCOUNT_SECURITY_KEY_OBJECT = "system/account-security/key.v1.json";
const ACCOUNT_SECURITY_KEY_VERSION = 1;
const ACCOUNT_SECURITY_KEY_BYTES = 32;
const ACCOUNT_SECURITY_KEY_SALT = "renewlet:account-security:v1";

export interface AccountSecurityKeyRing {
  totpSeed: CryptoKey;
  sharingCredential: CryptoKey;
  recoveryCode: CryptoKey;
  mfaTicket: CryptoKey;
  passkeyChallenge: CryptoKey;
  newszxcnIntegration: CryptoKey;
}

interface AccountSecurityKeyFile {
  version: number;
  key: string;
}

let accountSecurityKeyRingPromise: Promise<AccountSecurityKeyRing> | null = null;

export async function accountSecurityKeyRing(env: Env): Promise<AccountSecurityKeyRing> {
  if (!accountSecurityKeyRingPromise) {
    // 这里缓存的是安装级常量 key ring，不是请求/用户/session 状态；避免每次 OTP/Passkey 热路径都读 R2。
    accountSecurityKeyRingPromise = loadAccountSecurityKeyRing(env).catch((error: unknown) => {
      accountSecurityKeyRingPromise = null;
      throw error;
    });
  }
  return accountSecurityKeyRingPromise;
}

async function loadAccountSecurityKeyRing(env: Env): Promise<AccountSecurityKeyRing> {
  const master = await readOrCreateAccountSecurityMasterKey(env);
  const hkdfKey = await crypto.subtle.importKey("raw", arrayBufferFromBytes(master), "HKDF", false, ["deriveBits"]);
  // 同一个 cold-start 里只导入一次 HKDF master，四个用途靠 info 分域；Promise 缓存后热路径不再读 R2 或重复派生。
  const [totpSeedBytes, sharingCredentialBytes, recoveryCodeBytes, mfaTicketBytes, passkeyChallengeBytes, newszxcnBytes] = await Promise.all([
    deriveKeyBytes(hkdfKey, "totp-seed-aes-gcm"),
    deriveKeyBytes(hkdfKey, "sharing-credential-aes-gcm"),
    deriveKeyBytes(hkdfKey, "recovery-code-hmac"),
    deriveKeyBytes(hkdfKey, "mfa-ticket-hmac"),
    deriveKeyBytes(hkdfKey, "passkey-challenge-hmac"),
    deriveKeyBytes(hkdfKey, "newszxcn-integration-aes-gcm"),
  ]);
  return {
    totpSeed: await importAesGcmKey(totpSeedBytes),
    sharingCredential: await importAesGcmKey(sharingCredentialBytes),
    recoveryCode: await importHmacKey(recoveryCodeBytes),
    mfaTicket: await importHmacKey(mfaTicketBytes),
    passkeyChallenge: await importHmacKey(passkeyChallengeBytes),
    newszxcnIntegration: await importAesGcmKey(newszxcnBytes),
  };
}

async function readOrCreateAccountSecurityMasterKey(env: Env): Promise<Uint8Array> {
  const existing = await readAccountSecurityMasterKey(env);
  if (existing) return existing;
  const generated = crypto.getRandomValues(new Uint8Array(ACCOUNT_SECURITY_KEY_BYTES));
  const stored: AccountSecurityKeyFile = {
    version: ACCOUNT_SECURITY_KEY_VERSION,
    key: base64Url(generated),
  };
  // R2 条件写处理多个冷 isolate 同时首次启用账号安全；竞争失败后读取胜出的同一个安装级 key。
  const created = await env.ASSETS_BUCKET.put(ACCOUNT_SECURITY_KEY_OBJECT, `${JSON.stringify(stored)}\n`, {
    httpMetadata: { contentType: "application/json" },
    onlyIf: { etagDoesNotMatch: "*" },
  });
  if (created) return generated;
  const raced = await readAccountSecurityMasterKey(env);
  if (raced) return raced;
  throw new Error("account security key is unavailable");
}

async function readAccountSecurityMasterKey(env: Env): Promise<Uint8Array | null> {
  const object = await env.ASSETS_BUCKET.get(ACCOUNT_SECURITY_KEY_OBJECT);
  if (!object) return null;
  const raw = textDecoder.decode(await object.arrayBuffer());
  const parsed = JSON.parse(raw) as Partial<AccountSecurityKeyFile>;
  // R2 系统对象一旦损坏必须 fail closed；自动重建会让既有 TOTP/恢复码/Passkey challenge 全部不可验证。
  if (parsed.version !== ACCOUNT_SECURITY_KEY_VERSION || typeof parsed.key !== "string") {
    throw new Error("invalid account security key");
  }
  const key = base64UrlToArrayBuffer(parsed.key);
  if (key.byteLength !== ACCOUNT_SECURITY_KEY_BYTES) {
    throw new Error("invalid account security key");
  }
  return new Uint8Array(key);
}

async function importAesGcmKey(keyBytes: ArrayBuffer): Promise<CryptoKey> {
  return crypto.subtle.importKey("raw", keyBytes, { name: "AES-GCM" }, false, ["encrypt", "decrypt"]);
}

async function importHmacKey(keyBytes: ArrayBuffer): Promise<CryptoKey> {
  return crypto.subtle.importKey("raw", keyBytes, { name: "HMAC", hash: "SHA-256" }, false, ["sign"]);
}

async function deriveKeyBytes(key: CryptoKey, info: string): Promise<ArrayBuffer> {
  return crypto.subtle.deriveBits({
    name: "HKDF",
    hash: "SHA-256",
    salt: textEncoder.encode(ACCOUNT_SECURITY_KEY_SALT),
    info: textEncoder.encode(info),
  }, key, ACCOUNT_SECURITY_KEY_BYTES * 8);
}

export function resetAccountSecurityKeyRingForTest(): void {
  accountSecurityKeyRingPromise = null;
}

const textEncoder = new TextEncoder();
const textDecoder = new TextDecoder();
